### ISSUE TITLE
Attendtance 

### ISSUE THREAD

**Issue by Sameerabs** [Sameerabs/ForReview#1]
<!DOCTYPE html>
<html>

<head>
  <title>Attendance - HYT</title>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css">
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.16.0/umd/popper.min.js"></script>
  <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js"></script>
  <script src="https://unpkg.com/html5-qrcode"></script>

  <style>
    #qr-info {
      font-size: 50px;
      /* Increase font size */
    }

    #qr-info.success {
      color: green;
      /* Set color to green for successful input */
    }

    #qr-info.error {
      color: red;
      /* Set color to red for invalid input */
    }

    .row {
      margin: 20px 0;
    }
  </style>
</head>

<body>

  <div class="container">
    <div class="row">
      <div class="col-12 text-center">
        <h1>Attendance - HYT</h1>
      </div>
    </div>
    <div class="row">
      <div class="col-3">
      </div>
      <div class="col-6 text-center">
        <div class="input-group">
          <div class="input-group-prepend">
            <span class="input-group-text">Teacher Name:</span>
          </div>
          <select class="form-control" id="userName-drop">
    </select>
        </div>

        <div style="margin-top: 10px;"></div> 

        <div class="input-group">
          <div class="input-group-prepend">
            <span class="input-group-text">Session Time:</span>
          </div>
          <select class="form-control" id="session-Time-drop">
    </select>
        </div>
      </div>

      <div class="col-3">
      </div>
    </div>
    <div class="row">
      <div class="col-5 offset-1">
        <button id="scan-btn" class="btn btn-primary btn-block">Scan QR Code</button>
      </div>
      <div class="col-5">
        <button id="close-btn" class="btn btn-primary btn-block">Close Camera</button>
      </div>
    </div>
    <div class="row mt-3">
      <div class="col-12">
        <div id="qr-reader" style="width: 600px"></div>
      </div>
    </div>
    <div class="row mt-3">
      <div class="col-12">
        <label id="qr-info"></label>
      </div>
    </div>
    <div class="row mt-3">
      <div class="col-12">
        <label id="Code-Version" style="font-size: 6px;"> V11.0</label>
      </div>
    </div>
  </div>

  <script>
    var participants = [] ;
    var nciParticipants = [];
    var attendanceValList = [];
    var timeAllowedBeforeSession = 45;
    var timeAllowedAfterSession = 59;

  var qrInfoLabel = document.getElementById("qr-info");
  var sessionDropdown = document.getElementById("session-Time-drop");
  
  var scannedInfo = '';
  const sessionDropDownDefaultText = "Select Session Time";
  var sessionTime = sessionDropDownDefaultText ;
  var sessionID = '';
    var userNameDropDown = document.getElementById("userName-drop");
  const userNameDropDownDefaultText = "Select Teacher Name";
  var userName = userNameDropDownDefaultText ;

    // Initialize the scanner object
  var html5QrcodeScanner = new Html5QrcodeScanner(
	"qr-reader", { fps: 10, qrbox: 250 });

// Define a function to close the scanner and show the result in the label
function closeScanner(decodedText) {
    // Stop the scanner
    html5QrcodeScanner.stop();

    // Show the scanned text in the label
    $('#qr-info').text(decodedText);

    // Send the scanned data to server via Google Apps Script
    google.script.run.sendback(decodedText);
}

// Bind the click event to the Scan QR Code button
$('#scan-btn').click(function() {
  scanQRCode();
});

function scanQRCode() {
  let datenow = new Date();
  let allowedTime = new Date(sessionTime);
  let maxTime = new Date(sessionTime);
  allowedTime.setMinutes(allowedTime.getMinutes() - timeAllowedBeforeSession);
  maxTime.setMinutes(maxTime.getMinutes() + timeAllowedAfterSession);

  if (sessionTime == sessionDropDownDefaultText || userName == userNameDropDownDefaultText) {
    $('#qr-info').text('Please select Teacher Name / Session Time');
    qrInfoLabel.classList.add("error");
    qrInfoLabel.classList.remove("success");
  } else if (datenow >= allowedTime) {

    html5QrcodeScanner.render(onScanSuccess);
    $('#qr-info').text('');    

  } else {
    $('#qr-info').text('Too Early to mark the attendance');
    qrInfoLabel.classList.add("error");
    qrInfoLabel.classList.remove("success");
  }
}

function onScanSuccess(decodedText, decodedResult) {
  if (decodedText !== scannedInfo) {
  // Close the scanner
  html5QrcodeScanner.clear();
    getParticipantStatus(decodedText);
    scannedInfo = decodedText;
  }
}

// Bind the click event to the Close the camera QR Code button
$('#close-btn').click(function() {
    // Close the scanner
  html5QrcodeScanner.clear();

});

function getParticipantStatus(deCodedScannedData) {

  if (navigator.onLine) {
    let foundMatch = false; // Variable to track if ID is found
    let remarks = '';

    // Check if participants array is defined and not empty
    if (Array.isArray(participants) && participants.length > 0) {
      // Find the participant with matching participant ID
      let participant = participants.find(item => item[0] == deCodedScannedData);
      if (participant) {
        foundMatch = true;
        let participantID = participant[0];
        let participantName = participant[1];
        let bayName = participant[2];
        // Define other variables required for the code
        let i = 0; // Update with appropriate value

       // Rest of the code for ID match...
        markAttendance(participantID, enteredWithInTime(), i, sessionID, userName)
          .then(response => {
            // Handle success response
           // console.log(response);
            if (isParticipantNCI(participantID)) {
              remarks = '(' + participantID + ') ' + participantName + ' :\n' + bayName + ' | NCI';
              setRemarks(remarks, 'error');
             // console.log('NCI - ' + participantID);
            } else {
              remarks = '(' + participantID + ') ' + participantName + ' :\n' + bayName + '\n| Allowed';
              setRemarks(remarks, 'success');
              //console.log('No NCI - ' + participantID);
            }
          })
          .catch(error => {
            // Handle error response
            console.error(error);
            setRemarks('Error: Unable to mark attendance', 'error');
          });

        // Return the remarks.
        return remarks;
      } else {
        remarks = 'Error: Participant not found';
      }
    } else {
      remarks = 'Error: Participants list is empty or undefined';
    }
  } else {
    setRemarks('No internet connection', 'error');
    html5QrcodeScanner.clear();
  }

  return remarks;
}

// Function to set the remarks text and update CSS classes
function setRemarks(remarks, status) {
  $('#qr-info').text(remarks);
  qrInfoLabel.classList.add(status);
  qrInfoLabel.classList.remove(status === 'success' ? 'error' : 'success');
}

// Asynchronously mark attendance using google.script.run
async function markAttendance(participantID, enteredTime, index, sessionID, userName) {
  try {
    const response = await google.script.run.markAttendance(participantID, enteredTime, index, sessionID, userName);
    return response;
  } catch (error) {
    throw error;
  }
}

function isParticipantNCI(participantID) {
  return nciParticipants.includes(participantID);
}

function enteredWithInTime() {
  const now = new Date();
  const currentTime = now.getHours() * 60 + now.getMinutes();
  const sessionTimeValue = sessionTime.getHours() * 60 + sessionTime.getMinutes();

  return currentTime <= sessionTimeValue;
}

  function getParticipantData(list){
    participants = list.slice();
  }

  function getattendaceVolsList(vol_list){
    userNameDropDown.innerHTML = '';
    var defaultOption = document.createElement("option");
    defaultOption.text = userNameDropDownDefaultText;
    userNameDropDown.add(defaultOption);

    for (var j = 0; j < vol_list.length; j++) {
      var optionElement = document.createElement("option");
      optionElement.text = vol_list[j];    
      userNameDropDown.add(optionElement);
    }

  }

  function getnciParticipantData(list){
    nciParticipants = list.slice();
  }

  

  function getTheTimeBeforeSession(value){
    timeAllowedBeforeSession = value;
    //console.log('Allowed Time before is ' + timeAllowedBeforeSession);
  }

    function getTheTimeAfterSession(value){
    timeAllowedAfterSession = value;
   // console.log('Allowed Time after is ' + timeAllowedAfterSession);
  }

  function getSessionTimeList(session_List){
    sessionDropdown.innerHTML = '';
    var defaultOption = document.createElement("option");
    defaultOption.text = sessionDropDownDefaultText;
    sessionDropdown.add(defaultOption);

    for (var j = 0; j < session_List.length; j++) {
      var optionElement = document.createElement("option");
      optionElement.text = session_List[j];    
      sessionDropdown.add(optionElement);
    }
  }

sessionDropdown.addEventListener("change", function() { 
  let timeString = sessionDropdown.value;
  if(timeString !==sessionDropDownDefaultText){
    const [hours, minutes, seconds] = timeString.split(":");
    sessionTime = new Date();
    sessionTime.setHours(hours);
    sessionTime.setMinutes(minutes);
    sessionTime.setSeconds(seconds);
    sessionID = getCurrentDate()+ hours+minutes;

  }
  else{    
    html5QrcodeScanner.clear();
    sessionTime = sessionDropDownDefaultText;
  }
});

userNameDropDown.addEventListener("change", function() { 
  userName = userNameDropDown.value;
});

function getCurrentDate() {
  var currentDate = new Date();

  // Get day, month, and year
  var day = String(currentDate.getDate()).padStart(2, '0');
  var month = String(currentDate.getMonth() + 1).padStart(2, '0'); // Months are zero-based
  var year = String(currentDate.getFullYear());

  // Concatenate the components in DDMMYYYY format
  var formattedDate = year + month + day ;

  return formattedDate;
}

  function fillControlInformation(){
    google.script.run.withSuccessHandler(getParticipantData).getparticipants();     
    google.script.run.withSuccessHandler(getSessionTimeList).getSessionTime();    
    google.script.run.withSuccessHandler(getattendaceVolsList).getAttendaceVals();     
    google.script.run.withSuccessHandler(getTheTimeBeforeSession).getAllowedTimeBefore();     
    google.script.run.withSuccessHandler(getTheTimeAfterSession).getAllowedTimeAfter();     
    google.script.run.withSuccessHandler(getnciParticipantData).getNCIData();     

 }

  document.addEventListener('DOMContentLoaded', function() {
    fillControlInformation();
  });

  </script>

</body>

</html>
